tests/lang/s02_core/s02_procedures/t03_overload_core.nim

```diff
@@ -0,0 +1,185 @@
+discard """
+description: '''
+Specification of the core overloading resolution algorithm with examples
+'''
+
+"""
+
+## In a call `p(args)` the routine `p` that matches best is selected.
+## If multiple routines match equally well, the ambiguity is reported
+## during semantic analysis.
+
+block integer_literal:
+  proc impl(arg: uint32): string = "uint32"
+  proc impl(arg: uint8): string = "uint8"
+
+  doAssert impl(10u8) == "uint8", "Exact match always takes precedence"
+  doAssert impl(10u16) == "uint32", "When no explicit overload present, widen integer"
+
+block subrange_match:
+  ## Subrange - `arg` is a `range[T]` and `T` matches range of the
+  ## argment exactly or is a subrange of argument.
+  proc impl(arg: range[0 .. 10]): string = "0 .. 10"
+  proc impl(arg: range[11 .. 20]): string = "11 .. 20"
+
+  ## Allowed range of the integer values is take into account when overloading
+  ## is performed.
+  doAssert impl(range[0 .. 10](10)) == "0 .. 10"
+  doAssert impl(range[11 .. 20](11)) == "11 .. 20"
+
+  ## expression type can be a subrange of the expected procedure argument type
+  doAssert impl(range[0 .. 2](1)) == "0 .. 10"
+
+block subtype_match:
+  ## Subtype match - argument is a subtype of the procedure argument. Note that this
+  ## requirement is related to inheritance (because of subtyping), but does not
+  ## replace `method`'s functionality. It also works with `{.inheritable.}` types and
+  ## `object of RootObj` (in addition to `ref object of RootObj`) whereas `method`
+  ## requires `ref` to be used.
+  block inheritable_pragma:
+    ## `{.inheritable.}` provides subtyping capabilities
+    type
+      Base {.inheritable.} = object
+        field: int
+
+      Derived = object of Base
+        field2: int
+
+    proc impl(arg: Base) = discard
+
+    impl(Derived())
+    impl(Base())
+
+    ## Note that only embedded supertype part is passed to the procedure in this
+    ## case, and it is not safe to `cast` to a base type.
+    ##
+    ## ```
+    ## Derived[Sup = Base[field1: int], field2: int]
+    ##         ^^^^^^^^^^^^^^^^^^^^^^^
+    ##         Only this part of the object is passed to the procedure.
+    ## ```
+
+  block subtype_ref_ptr_var:
+    ## `ref` and `ptr` and `var` do not affect subtype relation - as a result they
+    ## can also be used to call a procedure that expects a supertype.
+    type
+      Base = object of RootObj
+        fbase: int
+
+      Derived = object of Base
+        fderived: int
+
+    proc impl(arg: ptr Base): string = "ptr"
+    proc impl(arg: ref Base): string = "ref"
+    proc impl(arg: var Base): string = arg = Base(fbase: 256); "var"
+
+    doAssert impl((ptr Derived)(nil)) == "ptr"
+    doAssert impl((ref Derived)(nil)) == "ref"
+
+    var derived = Derived(fderived: 128)
+
+    ## It is possible to modify part of the embedded supertype via `var` argument
```

Comment:
Maybe this should be moved to the "argument passing" part of the specification, but I'm not sure.

Reply:
argument passing styles/memory semantics makes sense, it'll be somewhat repetitive but that's not an issue.

Part of this shows which proc is called (belongs here), but the actual semantics of what's passed in and how mutations are allowed/do or don't show up is more about passing.